File: isma-compiler/hsm-core/src/main/java/ru/nstu/isma/core/hsm/linear/LinearEquationsSolver.java
package ru.nstu.isma.core.hsm.linear;


import ru.nstu.isma.print.MatrixPrint;

/**
 * Created by Дмитрий Достовалов
 * on 13.03.2015.
 */

public class LinearEquationsSolver {
    double[][] LU;  // LU-разложение матрицы A для задачи Ax=b.
    int[] kd;       // При LU-разложении выполняется выбор ведущего элемента,
    // т.е. перестановки строк. Здесь запоминаются перестановки.
    int size;

    public LinearEquationsSolver(double[][] A) {
        MatrixPrint.print("input", A);
        size = A.length;
        LU = new double[size][size];
        for (int i = 0; i < size; i++) {
            if (A[i].length != size) throw new RuntimeException("Different matrix A size!");
            for (int j = 0; j < size; j++) {
                LU[i][j] = A[i][j];
            }
        }
        kd = new int[size];
        Dec();      // Выполняется LU-разложение
    }

    private void Dec() {
        int m;
        double t;

        kd[size - 1] = 1;
        if (size > 1) {
            for (int j = 0; j < size - 1; j++) {
                m = j;
                for (int i = j + 1; i < size; i++)
                    if (Math.abs(LU[i][j]) > Math.abs(LU[m][j])) m = i;
                kd[j] = m;
                t = LU[m][j];
                if (m != j) {
                    kd[size - 1] = -kd[size - 1];
                    LU[m][j] = LU[j][j];
                    LU[j][j] = t;
                }
                if (Math.abs(t) < 1.0e-20) {
                    kd[size - 1] = 0;
                    throw new RuntimeException("Matrix A is singular!");
                }
                t = 1.0 / t;
                for (int i = j + 1; i < size; i++)
                    LU[i][j] = -LU[i][j] * t;
                for (int i = j + 1; i < size; i++) {
                    t = LU[m][i];
                    LU[m][i] = LU[j][i];
                    LU[j][i] = t;
                    if (Math.abs(t) > 1.0e-20)
                        for (int k = j + 1; k < size; k++)
                            LU[k][i] = LU[k][i] + LU[k][j] * t;
                }
            }
        }
        if (Math.abs(LU[size - 1][size - 1]) < 1.0e-20) {
            kd[size - 1] = 0;
            throw new RuntimeException("Matrix A is singular!");
        }
    }

    public double[] solve(double[] b) {
        MatrixPrint.print("LU", LU, b);
        if (b.length != size) throw new RuntimeException("Different size!");

        double[] x = new double[size];
        for (int i = 0; i < size; i++)
            x[i] = b[i];

        double t;
        int m;

        if (size > 1) {
            for (int j = 0; j < size - 1; j++) {
                m = kd[j];
                t = x[m];
                x[m] = x[j];
                x[j] = t;
                for (int i = j + 1; i < size; i++)
                    x[i] = x[i] + LU[i][j] * t;
            }
            for (int j = size - 1; j > 0; j--) {
                x[j] = x[j] / LU[j][j];
                t = -x[j];
                for (int i = 0; i < j; i++)
                    x[i] = x[i] + LU[i][j] * t;
            }
        }
        x[0] = x[0] / LU[0][0];

        return x;
    }
}

File: isma-next-hsm/src/main/kotlin/ru/nstu/isma/hsm/linear/LinearEquationsSolver.kt
package ru.nstu.isma.hsm.linear

import ru.nstu.isma.print.MatrixPrint.print
import kotlin.math.abs

/**
 * Created by Дмитрий Достовалов
 * on 13.03.2015.
 */
class LinearEquationsSolver(A: Array<DoubleArray>) {
    var LU // LU-разложение матрицы A для задачи Ax=b.
            : Array<DoubleArray>
    var kd // При LU-разложении выполняется выбор ведущего элемента,
            : IntArray

    // т.е. перестановки строк. Здесь запоминаются перестановки.
    var size: Int
    private fun Dec() {
        var m: Int
        var t: Double
        kd[size - 1] = 1
        if (size > 1) {
            for (j in 0 until size - 1) {
                m = j
                for (i in j + 1 until size) if (abs(LU[i][j]) > abs(LU[m][j])) m = i
                kd[j] = m
                t = LU[m][j]
                if (m != j) {
                    kd[size - 1] = -kd[size - 1]
                    LU[m][j] = LU[j][j]
                    LU[j][j] = t
                }
                if (abs(t) < 1.0e-20) {
                    kd[size - 1] = 0
                    throw RuntimeException("Matrix A is singular!")
                }
                t = 1.0 / t
                for (i in j + 1 until size) LU[i][j] = -LU[i][j] * t
                for (i in j + 1 until size) {
                    t = LU[m][i]
                    LU[m][i] = LU[j][i]
                    LU[j][i] = t
                    if (abs(t) > 1.0e-20) for (k in j + 1 until size) LU[k][i] = LU[k][i] + LU[k][j] * t
                }
            }
        }
        if (abs(LU[size - 1][size - 1]) < 1.0e-20) {
            kd[size - 1] = 0
            throw RuntimeException("Matrix A is singular!")
        }
    }

    fun solve(b: DoubleArray?): DoubleArray {
        print("LU", LU, b!!)
        if (b.size != size) throw RuntimeException("Different size!")
        val x = DoubleArray(size)
        for (i in 0 until size) x[i] = b[i]
        var t: Double
        var m: Int
        if (size > 1) {
            for (j in 0 until size - 1) {
                m = kd[j]
                t = x[m]
                x[m] = x[j]
                x[j] = t
                for (i in j + 1 until size) x[i] = x[i] + LU[i][j] * t
            }
            for (j in size - 1 downTo 1) {
                x[j] = x[j] / LU[j][j]
                t = -x[j]
                for (i in 0 until j) x[i] = x[i] + LU[i][j] * t
            }
        }
        x[0] = x[0] / LU[0][0]
        return x
    }

    init {
        print("input", A)
        size = A.size
        LU = Array(size) { DoubleArray(size) }
        for (i in 0 until size) {
            if (A[i].size != size) throw RuntimeException("Different matrix A size!")
            for (j in 0 until size) {
                LU[i][j] = A[i][j]
            }
        }
        kd = IntArray(size)
        Dec() // Выполняется LU-разложение
    }
}